src/main/java/BasicMAPF/Solvers/AStar/SingleAgentAStarSIPP_Solver.java

```diff
@@ -0,0 +1,311 @@
+            timesList.add(timeLocation.getTime());
+        }
+
+        HashMap<I_Location, ArrayList<Interval>> intervalMap = new HashMap<>();
+        for (I_Location location : timeIntervals.keySet()) {
+            ArrayList<Integer> timestamps = timeIntervals.get(location);
+
+            // Convert the timestamps to intervals using timestampsToIntervals function
+            ArrayList<Interval> intervals = timestampsToFreeIntervals(timestamps);
+
+            // Update the output map with the intervals for the current location
+            intervalMap.put(location, intervals);
+        }
+
+        // Add [0, inf] interval for locations in allGraphLocations that are not in intervalMap
+        for (I_Coordinate coordinate : allGraphLocations.keySet()) {
+            I_Location location = allGraphLocations.get(coordinate);
+            if (!intervalMap.containsKey(location)) {
+                ArrayList<Interval> infInterval = new ArrayList<>();
+                infInterval.add(new Interval(0, Integer.MAX_VALUE));
+                intervalMap.put(location, infInterval);
+            }
+        }
+
+        return intervalMap;
+    }
+
+    private static ArrayList<Interval> timestampsToFreeIntervals(ArrayList<Integer> timestamps) {
+        // Sort timestamps in ascending order
+        Collections.sort(timestamps);
+
+        ArrayList<Interval> freeIntervals = new ArrayList<>();
+
+        // Add the first interval if it starts from more than 0
+        if (timestamps.get(0) > 0) {
+            freeIntervals.add(new Interval(0, timestamps.get(0) - 1));
+        }
+
+        // Iterate through timestamps to find free intervals
+        for (int i = 1; i < timestamps.size(); i++) {
+            if (timestamps.get(i) > timestamps.get(i - 1) + 1) {
+                freeIntervals.add(new Interval(timestamps.get(i - 1) + 1, timestamps.get(i) - 1));
+            }
+        }
+
+        // Add the last interval extending to positive infinity
+        freeIntervals.add(new Interval(timestamps.get(timestamps.size() - 1) + 1, Integer.MAX_VALUE));
+
+        return freeIntervals;
+    }
+
+    public class AStarSIPPState extends AStarState {
+        private final Interval timeInterval;
+
+        public AStarSIPPState(Move move, AStarSIPPState prev, int g, int conflicts, Interval timeInterval, boolean isMoveToTargetLocation) {
+            super(move, prev, g, conflicts, isMoveToTargetLocation);
+            this.timeInterval = timeInterval;
+
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (!(o instanceof AStarSIPPState that)) return false;
+
+            if (timeInterval.start != that.timeInterval.start) return false;
+            assert move != null;
+            assert that.move != null;
+            return move.currLocation.equals(that.move.currLocation);
+        }
```

Comment:
Could you add a short comment (javadoc) explaining why only these fields must be checked?